lib/Sema/TypeCheckType.cpp

```diff
@@ -1943,7 +1938,34 @@ Type ResolveTypeRequest::evaluate(Evaluator &evaluator,
   if (validateAutoClosureAttributeUse(ctx.Diags, TyR, result, options))
     return ErrorType::get(ctx);
 
-  return result;
+  // Diagnose an attempt to use a placeholder at the top level.
+  if (result->getCanonicalType()->is<PlaceholderType>() &&
```

Comment:
@rintaro Updated this portion to check the canonical type! Turned out to be a bit more annoying than expected since a `PlaceholderTypeRepr` isn't always transformed into a `PlaceholderType`—the transformation is based on a client-specified function.

Now, we resolve every `PlaceholderTypeRepr` into a `PlaceholderType`, then after diagnosing top-level placeholders we proceed with the transformation step below. Since the logic here has changed nontrivially would love if you could give it another look!